Add DeepFloyd If deployment blog post (#1320)
huggingface/blog: deploy-deepfloydif-using-bentoml.md

---
title: "Deploying Hugging Face Models with BentoML: DeepFloyd IF in Action" 
thumbnail: /blog/assets/deploy-deepfloydif-using-bentoml/thumbnail.png
authors:

> - user: Sherlockk
> - user: larme

Reviewer: ```suggestion
- user: Sherlockk
  guest: true
- user: larme
  guest: true
```

---

# Deploying Hugging Face Models with BentoML: DeepFloyd IF in Action

<!-- {blog_metadata} -->
<!-- {authors} -->

> Hugging Face provides a centralized Model Hub that allows you to upload, share, and deploy your models with ease. It saves developers the time and computational resources required to train models from scratch. However, deploying models in a real-world production environment or in a cloud-native way can still present challenges.

Reviewer: Let's maybe call it the Hub platform as it's not specific to models only. Hub platform allows for easy sharing of datasets and ML application demos too.

This is where BentoML comes into the picture. BentoML is an open-source platform for machine learning model serving and deployment. It is a unified framework for building, shipping, and scaling production-ready AI applications incorporating traditional ML models, pre-trained AI models, Generative models, and Large Language Models. Here is how you use the BentoML framework from a high-level perspective:

1. **Define a model**: Before you can use BentoML, you need a machine learning model (or multiple models). This model can be trained using a machine learning library such as TensorFlow and PyTorch.

> 2. **Save the model**: Once you have a trained model, save it to the BentoML local Model Store, which is used for managing all your trained models locally as well as accessing them for serving.

Reviewer: So, the resultant image will _always_ have the model / artifact files?

Author: Yes, correct.

5. **Deploy the Bento**: Once the Bento is ready, you can containerize the Bento to create a Docker image and run it on Kubernetes. Alternatively, deploy the Bento directly to Yatai, an open-source, end-to-end solution for automating and running machine learning deployments on Kubernetes at scale.

If you seek a fully managed service for building and operating AI applications, consider deploying Bentos through BentoCloud. All [Hugging Face Diffuser models](https://huggingface.co/docs/diffusers/index) and pipelines can be seamlessly integrated with BentoML and shipped to production via BentoCloud, which enables models to run on the most suitable hardware with independent scalability based on usage.

In this blog post, we will demonstrate how to integrate [DeepFloyd IF](https://huggingface.co/docs/diffusers/api/pipelines/if) with BentoML by following the above workflow. DeepFloyd IF is a state-of-the-art, open-source text-to-image model. It stands apart from latent diffusion models like Stable Diffusion due to its distinct operational strategy and architecture.

> 

Reviewer: I'd also add a ToC here.

> DeepFloyd delivers a high degree of photorealism and sophisticated language understanding. Unlike Stable Diffusion, DeepFloyd IF works directly in pixel space, leveraging a modular structure that encompasses a frozen text encoder and three cascaded pixel diffusion modules. Each module plays a unique role in the process: Stage 1 is responsible for the creation of a base 64x64 px image, which is then progressively upscaled to 1024x1024 px across Stage 2 and Stage 3. Another critical aspect of DeepFloyd IF’s uniqueness is its integration of a Large Language Model (T5-XXL-1.1) to encode prompts, which offers superior understanding of complex prompts. For more information, see this [Stability AI blog post about DeepFloyd IF](https://stability.ai/blog/deepfloyd-if-text-to-image-model).

Reviewer: I'd add a header to this section, maybe something like "Brief intro to IF".

> ## Creating a DeepFloyd IF application with BentoML

Reviewer: Maybe discuss why deploying IF is particularly challenging?

Author: It is actually not just about deploying IF. BentoML works for all ML models. We use IF here just as an example to demo how BentoML works. Using the BentoML framework provides developers a unified way to deploy their AI apps in production. BentoML helps developers and in-house data science teams bridge the gap between model creation and production deployment. Perhaps I can add some explanations for it at the beginning to clarify this.

Reviewer: I think that is clear. But we could have shown this with a `StableDiffusionPipeline` too. But I don't think it's particularly challenging to deploy. The `IFPipeline` is because of its computational requirements. Perhaps, it's worth clarifying that part and how we leverage the superpowers of BentoML to help ease that process.

Does it make more sense now?

Author: Thanks. I added some explanations about BentoML's ability to allow users manage resources for IF deployment.

[This GitHub repository](https://github.com/bentoml/IF-multi-GPUs-demo) stores all necessary files for this project. To run this project locally, make sure you have the following:

- Python 3.8+
- `pip` installed

> - At least 2x16GB VRAM GPU or 1x40 VRAM GPU. For this project, we used a machine of type `n1-standard-16` from Google Cloud plus 64 GB of RAM and 2 NVIDIA T4 GPUs.

Reviewer: But it's also possible to run IF on a single T4: https://huggingface.co/blog/if. 

I'd maybe make a note about this and then warn the readers that even though the support is there, it's not really recommended for production-grade serving. 

WDYT?

Author: I am ok to this. Perhaps add a sentence like this: `Note that while it is possible to run IF on a single T4, it is not recommended for production-grade serving`. @larme Let me know if you have any comments on it.

Once the prerequisites are met, clone the project repository to your local machine and navigate to the target directory.

```
git clone https://github.com/bentoml/IF-multi-GPUs-demo.git
cd IF-multi-GPUs-demo
```

Before building the application, let’s briefly explore the key files within this directory:

> - `import_models.py`: Defines the models for each Stage. You use this file to download all the models to your local machine so that you can package them into a single Bento.

Reviewer: ```suggestion
- `import_models.py`: Defines the models for each stage of the [`IFPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/if). You use this file to download all the models to your local machine so that you can package them into a single Bento.
```

---
title: "Deploying Hugging Face Models with BentoML: DeepFloyd IF in Action" 
thumbnail: /blog/assets/deploy-deepfloydif-using-bentoml/thumbnail.png
authors:
- user: Sherlockk
  guest: true
- user: larme
  guest: true
---

# Deploying Hugging Face Models with BentoML: DeepFloyd IF in Action

<!-- {blog_metadata} -->
<!-- {authors} -->

Hugging Face provides a Hub platform that allows you to upload, share, and deploy your models with ease. It saves developers the time and computational resources required to train models from scratch. However, deploying models in a real-world production environment or in a cloud-native way can still present challenges.

> This is where BentoML comes into the picture. BentoML is an open-source platform for machine learning model serving and deployment. It is a unified framework for building, shipping, and scaling production-ready AI applications incorporating traditional ML models, pre-trained AI models, Generative models, and Large Language Models. Here is how you use the BentoML framework from a high-level perspective:

Reviewer 2: ```suggestion
This is where BentoML comes into the picture. BentoML is an open-source platform for machine learning model serving and deployment. It is a unified framework for building, shipping, and scaling production-ready AI applications incorporating traditional, pre-trained, and generative models as well as Large Language Models. Here is how you use the BentoML framework from a high-level perspective:
```

2. **Save the model**: Once you have a trained model, save it to the BentoML local Model Store, which is used for managing all your trained models locally as well as accessing them for serving.
3. **Create a BentoML Service**: You create a `service.py` file to wrap the model and define the serving logic. It specifies [Runners](https://docs.bentoml.org/en/latest/concepts/runner.html) for models to run model inference at scale and exposes APIs to define how to process inputs and outputs.
4. **Build a Bento**: By creating a configuration YAML file, you package all the models and the [Service](https://docs.bentoml.org/en/latest/concepts/service.html) into a [Bento](https://docs.bentoml.org/en/latest/concepts/bento.html), a deployable artifact containing all the code and dependencies.
5. **Deploy the Bento**: Once the Bento is ready, you can containerize the Bento to create a Docker image and run it on Kubernetes. Alternatively, deploy the Bento directly to Yatai, an open-source, end-to-end solution for automating and running machine learning deployments on Kubernetes at scale.

> If you seek a fully managed service for building and operating AI applications, consider deploying Bentos through BentoCloud. All [Hugging Face Diffuser models](https://huggingface.co/docs/diffusers/index) and pipelines can be seamlessly integrated with BentoML and shipped to production via BentoCloud, which enables models to run on the most suitable hardware with independent scalability based on usage.

Reviewer 3: ```suggestion
```

cd IF-multi-GPUs-demo
```

Before building the application, let’s briefly explore the key files within this directory:

- `import_models.py`: Defines the models for each stage of the [`IFPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/if). You use this file to download all the models to your local machine so that you can package them into a single Bento.
- `requirements.txt`: Defines all the packages and dependencies required for this project.
- `service.py`: Defines a BentoML Service, which contains three Runners created using the `to_runner` method and exposes an API for generating images. The API takes a JSON object as input (i.e. prompts and negative prompts) and returns an image as output by using a sequence of models.
- `start-server.py`: Starts a BentoML HTTP server through the Service defined in `service.py` and creates a Gradio web interface for users to enter prompts to generate images.
- `bentofile.yaml`: Defines the metadata of the Bento to be built, including the Service, Python packages, and models.

We recommend you create a Virtual Environment for dependency isolation. For example, run the following command to activate `myenv`:

> ```

Reviewer 2: nit for nice formatting, also for other blocks

```suggestion
```bash
```

Building BentoML service "deepfloyd-if:6ufnybq3vwszgnry" from build context "/Users/xxx/Documents/github/IF-multi-GPUs-demo".
Packing model "sd-upscaler:bb2ckpa3uoypynry"
Packing model "if-stage1:v1.0"
Packing model "if-stage2:v1.0"
Locking PyPI package versions.

██████╗░███████╗███╗░░██╗████████╗░█████╗░███╗░░░███╗██╗░░░░░
██╔══██╗██╔════╝████╗░██║╚══██╔══╝██╔══██╗████╗░████║██║░░░░░
██████╦╝█████╗░░██╔██╗██║░░░██║░░░██║░░██║██╔████╔██║██║░░░░░
██╔══██╗██╔══╝░░██║╚████║░░░██║░░░██║░░██║██║╚██╔╝██║██║░░░░░
██████╦╝███████╗██║░╚███║░░░██║░░░╚█████╔╝██║░╚═╝░██║███████╗
╚═════╝░╚══════╝╚═╝░░╚══╝░░░╚═╝░░░░╚════╝░╚═╝░░░░░╚═╝╚══════╝

Successfully built Bento(tag="deepfloyd-if:6ufnybq3vwszgnry").
```

View the Bento in the local Bento Store.

```
$ bentoml list

Tag                               Size       Creation Time
deepfloyd-if:6ufnybq3vwszgnry     49.25 GiB  2023-07-06 11:34:52
```

The Bento is now ready for serving in production.

```
bentoml serve deepfloyd-if:6ufnybq3vwszgnry
```

To deploy the Bento in a more cloud-native way, generate a Docker image by running the following command:

```
bentoml containerize deepfloyd-if:6ufnybq3vwszgnry
```

> You can then deploy the model on Kubernetes. Alternatively, push the Bento to Yatai or BentoCloud. For more information, see [Deploying Bentos](https://docs.bentoml.org/en/latest/concepts/deploy.html).

Reviewer 3: ```suggestion
You can then deploy the model on Kubernetes.
```
